src/Eloquent/CacheTrait.php

```diff
@@ -19,7 +19,17 @@
      */
     public static function bootCacheTrait() {
         static::saving(function (Model $model) {
-            $dirty = $model->getDirtyCached();
+            $props = $model->getCachedProperties();
+            $casts = $model->getCasts();
+
+            $dirty = [];
+
+            foreach ($props as $prop) {
+                $inner = $model->$prop;
+                $val = (array_key_exists($prop, $casts) ? self::nullCastAttribute($prop, $inner, $casts) : $inner);
+
+                $dirty[$prop] = $val;
+            }
```

Comment:
Instead of allways storing all dirty properties, just add them to dirty when hydrating cached properties.